src/packbuilder.rs

```diff
@@ -0,0 +1,377 @@
+        RawPackBuilder { raw: ptr, _marker: marker::PhantomData }
+    }
+    fn raw(&self) -> *mut raw::git_packbuilder { self.raw }
+}
+
+impl<'repo> Deref for PackBuilder<'repo> {
+    type Target = RawPackBuilder<'repo>;
+    fn deref(&self) -> &RawPackBuilder<'repo> { &self.inner }
+}
+
+impl<'repo> DerefMut for PackBuilder<'repo> {
+    fn deref_mut(&mut self) -> &mut RawPackBuilder<'repo> { &mut self.inner }
+}
+
+impl<'repo> Deref for ThreadedPackBuilder<'repo> {
+    type Target = RawPackBuilder<'repo>;
+    fn deref(&self) -> &RawPackBuilder<'repo> { &self.inner }
+}
+
+impl<'repo> DerefMut for ThreadedPackBuilder<'repo> {
+    fn deref_mut(&mut self) -> &mut RawPackBuilder<'repo> { &mut self.inner }
+}
+
+impl<'repo> Deref for NonThreadedPackBuilder<'repo> {
+    type Target = RawPackBuilder<'repo>;
+    fn deref(&self) -> &RawPackBuilder<'repo> { &self.inner }
+}
+
+impl<'repo> DerefMut for NonThreadedPackBuilder<'repo> {
+    fn deref_mut(&mut self) -> &mut RawPackBuilder<'repo> { &mut self.inner }
+}
+
+impl<'repo> Drop for RawPackBuilder<'repo> {
+    fn drop(&mut self) {
+        unsafe { raw::git_packbuilder_free(self.raw) }
+    }
+}
+
+impl Binding for Stage {
+    type Raw = raw::git_packbuilder_stage_t;
+    unsafe fn from_raw(raw: raw::git_packbuilder_stage_t) -> Stage {
+        match raw {
+            raw::GIT_PACKBUILDER_ADDING_OBJECTS => Stage::AddingObjects,
+            raw::GIT_PACKBUILDER_DELTAFICATION => Stage::Deltafication,
+            _ => panic!("Unknown git diff binary kind"),
+        }
+    }
+    fn raw(&self) -> raw::git_packbuilder_stage_t {
+        match *self {
+            Stage::AddingObjects => raw::GIT_PACKBUILDER_ADDING_OBJECTS,
+            Stage::Deltafication => raw::GIT_PACKBUILDER_DELTAFICATION,
+        }
+    }
+}
+
+extern fn foreach_c(buf: *const c_void, size: size_t,
+                    data: *mut c_void) -> c_int {
+    unsafe {
+        let buf = slice::from_raw_parts(buf as *const u8, size as usize);
+
+        let r = panic::wrap(|| {
+            let data = data as *mut &mut ForEachCb;
+            (*data)(buf)
+        });
+        if r == Some(true) {0} else {-1}
+    }
+}
+
+extern fn progress_c(stage: raw::git_packbuilder_stage_t, current: c_uint,
+                     total: c_uint, data: *mut c_void) -> c_int {
+    unsafe {
+        let stage = Binding::from_raw(stage);
+
+        let data = data as *mut ProgressCbPayload;
+        let r = panic::wrap(|| {
+            if let Some(ref mut callback) = (*data).callback {
+                callback(stage, current, total)
+            } else {
+                false
+            }
+        });
+        if r.is_none() {
+            // git can attempt to call the progress callback
+            // again even if it returns false.
+            // If the callback panicked then we probably want
+            // to ensure it doesn't get called again.
+            (*data).callback.take();
```

Comment:
I probably should have looked more into how that was implemented, just assumed it couldn't be doing anything like that by the method signature :disappointed:.

How will the `thread_local` interact with threads spawned by `libgit2`? If it does remain thread local then that implementation won't work for this case; the callback could panic on one thread then get called again on another thread which won't have its `LAST_ERROR` set.